gobblin-yarn/src/main/java/gobblin/yarn/GobblinApplicationMaster.java

```diff
@@ -0,0 +1,452 @@
+    shutdownRequest.setMsgSubType(HelixMessageSubTypes.WORK_UNIT_RUNNER_SHUTDOWN.toString());
+    shutdownRequest.setMsgState(Message.MessageState.NEW);
+
+    int messagesSent = this.helixManager.getMessagingService().send(criteria, shutdownRequest);
+    if (messagesSent == 0) {
+      LOGGER.error(String.format("Failed to send the %s message to the participants", shutdownRequest.getMsgSubType()));
+    }
+  }
+
+  /**
+   * A custom implementation of {@link LiveInstanceChangeListener}.
+   */
+  private static class GobblinLiveInstanceChangeListener implements LiveInstanceChangeListener {
+
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      for (LiveInstance liveInstance : liveInstances) {
+        LOGGER.info("Live Helix participant instance: " + liveInstance.getInstanceName());
+      }
+    }
+  }
+
+  /**
+   * A custom {@link MessageHandlerFactory} for {@link MessageHandler}s that handle messages of type
+   * {@link org.apache.helix.model.Message.MessageType#SHUTDOWN} for shutting down the controller.
+   */
+  private class ControllerShutdownMessageHandlerFactory implements MessageHandlerFactory {
+
+    @Override
+    public MessageHandler createHandler(Message message, NotificationContext context) {
+      return new ControllerShutdownMessageHandler(message, context);
+    }
+
+    @Override
+    public String getMessageType() {
+      return Message.MessageType.SHUTDOWN.toString();
+    }
+
+    @Override
+    public void reset() {
+
+    }
+
+    /**
+     * A custom {@link MessageHandler} for handling messages of sub type
+     * {@link HelixMessageSubTypes#APPLICATION_MASTER_SHUTDOWN}.
+     */
+    private class ControllerShutdownMessageHandler extends MessageHandler {
+
+      public ControllerShutdownMessageHandler(Message message, NotificationContext context) {
+        super(message, context);
+      }
+
+      @Override
+      public HelixTaskResult handleMessage() throws InterruptedException {
+        String messageSubType = this._message.getMsgSubType();
+
+        HelixTaskResult result = new HelixTaskResult();
+
+        if (messageSubType.equalsIgnoreCase(HelixMessageSubTypes.APPLICATION_MASTER_SHUTDOWN.toString())) {
+          if (stopInProgress) {
+            result.setSuccess(true);
+            return result;
+          }
+
+          LOGGER.info("Handling message " + HelixMessageSubTypes.APPLICATION_MASTER_SHUTDOWN.toString());
+
+          ScheduledExecutorService shutdownMessageHandlingCompletionWatcher =
+              MoreExecutors.getExitingScheduledExecutorService(new ScheduledThreadPoolExecutor(1));
+
+          // Schedule the task for watching on the removal of the shutdown message, which indicates that
+          // the message has been successfully processed and it's safe to disconnect the HelixManager.
+          shutdownMessageHandlingCompletionWatcher.scheduleAtFixedRate(new Runnable() {
```

Comment:
Sure. But I still feel we should switch to Akka for messaging sometime in the future as Helix's way of handling messaging and the use of ZooKeeper for message queues are not so sounding.